src/jsd/actuator_fsm_helpers.cc

```diff
@@ -540,7 +578,36 @@ fastcat::FaultType fastcat::Actuator::ProcessCS()
     return ALL_DEVICE_FAULT;
   }
 
-  if ((cycle_mono_time_ - last_transition_time_) > 5 * loop_period_) {
+  switch (actuator_sms_) {
+    case ACTUATOR_SMS_CSP: {
+      // account for updated position offset
+      double dt = fmax(
+          (state_->time - last_device_cmd_.actuator_csp_cmd.request_time), 0.0);
+      double offset_target_position =
```

Comment:
I am not a fan of this. Currently, this is an interface to the drive's Cyclic Synchronous Position mode. If this is added as the default behavior, Fastcat loses its direct interface to DS-402's Cyclic Synchronous Position mode.
I believe that, at the very least, this behavior should be optional, not the default, or even better, its own state/mode in Actuator's state machine.
A couple other things to take into account. First, I believe the drive itself performs some form of interpolation in between reception of CSP commands. Second, the drive supports an Interpolated Position mode when using CANOpen, in addition to the cyclic synchronous and profiled modes. This further leads me to think the behavior implemented here should be its own state/mode.

Reply:
We discovered that when fastcat receives CSP setpoints asynchronously from another module, the latency between when the motion profile is generated in the calling module (e.g. arm) and when it is sent to the Elmo (e.g. fcat/fastcat) can cause irregularities in the motion profile. Additionally, if the two modules are set up to work synchronously, one missed message can cause high current response and instability.

Previously if fastcat did not receive a CSP set point synchronously in one loop period, it would send the same CSP setpoint to the elmo again (effectively a zero-order hold). When the CSP setpoint has a feed-forward velocity term, it is not consistent to send the same CSP setpoint again. 

This code update updates the position offset based on the previously provided CSP command, thereby allowing fastcat to run at a different loop rate than a module generating motion profiles, and makes it robust to missed messages. 

We have tested this on the clark testbed with arm running at 250/256Hz, and fcat/fastcat running at 250/500/1000Hz and have seen much improved current response vs without interpolation. There is precedent for this interpolation approach as it was used broadly across many m2020 testbeds in the previous ethercat driver.

Reply:
I will post the results of the testing here when this is ready to move out of draft status.